### ISSUE TITLE
Crash from diff_context_line__pattern_match

### ISSUE THREAD

**Issue by kevinsawicki** [libgit2/libgit2#2586]
Reported over on https://github.com/atom/atom/issues/3627 with the full stack here: https://gist.github.com/matthewpennell/260f5f21c1109aab78d2

This is using `0.21.0`

```
Crashed Thread:  0  CrRendererMain  Dispatch queue: com.apple.main-thread
 
Exception Type:  EXC_CRASH (SIGABRT)
Exception Codes: 0x0000000000000000, 0x0000000000000000
 
Application Specific Information:
[72980] stack overflow
 
Thread 0 Crashed:: CrRendererMain  Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	0x00007fff83bc2866 __pthread_kill + 10
1   libsystem_pthread.dylib       	0x00007fff8583735c pthread_kill + 92
2   libsystem_c.dylib             	0x00007fff8d131bba __abort + 145
3   libsystem_c.dylib             	0x00007fff8d13246d __stack_chk_fail + 196
4   git.node                      	0x000000011ad5bcda diff_context_line__pattern_match + 251
5   git.node                      	0x000000011ad5bbac diff_context_find + 91
6   git.node                      	0x000000011ada8b72 get_func_line + 155
7   git.node                      	0x000000011ada885b xdl_emit_diff + 703
8   git.node                      	0x000000011ada8522 xdl_diff + 168
9   git.node                      	0x000000011ad61f6b git_xdiff + 299
10  git.node                      	0x000000011ad5d304 diff_patch_generate + 545
11  git.node                      	0x000000011ad5daa8 git_patch_from_diff + 413
```

**Comment by kevinsawicki** [comment]
@arrbee just got another report of this in https://github.com/atom/atom/issues/5485

This comment seems to have more precise reproduction steps, https://github.com/atom/atom/issues/3627#issuecomment-65845894

**Comment by kevinsawicki** [comment]
Also @ChrisBrough reports it no longer crashing after removing this line from the config:

```
[diff ...]
     xfuncname = ^[-+@a-zA-Z_].*$
```

**Comment by arrbee** [comment]
Hmm. Well, I tried rewriting one of the `xfuncname` tests in the test suite to use `^[-+@a-zA-Z_].*$` in the config file as test data and it does not crash. All that means is that the pattern alone in not sufficient to cause the bug - there must be some combination of source file data which triggers the problem when used in conjunction with the pattern.

The existing tests are not thorough with regards to the types of diffs in the file, so it is possible that there is some pattern of diff which results in the context line scan attempting to apply the pattern to invalid data (e.g. maybe if just the first byte is deleted or just the first byte inserted or just the last byte modified, etc). I think the next step is to make a libgit2 test that uses the `xfuncname = ^[-+@a-zA-Z_].*$` pattern (thank you @ChrisBrough for that) along with a more exhaustive set of programmatically generated diffs to see if there is some diff data pattern that induces this problem.

**Comment by ChrisBrough** [comment]
After looking at it further, it's specifically crashing on `Objective-C++` using standard C++ syntax. Maybe try the following the `[diff "objcpp"]` tag and `.mm` file with some class member functions.

**Comment by mak-ableton** [comment]
I'm encountering this crash for modified cucumber `*.feature` files. In the repo in question, the `.gitattributes` entry for those is `*.feature text diff=feature`. The corresponding entry in `~/.gitconfig` is 
```
[diff "feature"]
  xfuncname = ((^Feature:|Scenario:|Scenario Outline:).*)
```
Removing either the `diff` entry in `.gitattributes` or the `xfuncname` entry in `~/.gitconfig` avoids the crash, so I'm happy to have a workaround. Maybe this info helps in finding a proper fix.

**Comment by toddmazierski** [comment]
Hi, all!

A little bit of background: I'm an Atom user, and suffer from frequent crashes. About once a day a file I'm working on will crash, and keeps crashing every time it's opened.

The crash reports are similar to the one posted in this ticket: a `stack overflow` with `diff_context_line__pattern_match` in the trace. Running `git stash` or disabling Atom's "Git Diff" package allows me to work around the problem. _Unlike_ a few posters above, I do not have any `[diff …]` directives in my Git configuration. If I'm able to distill it down to a reproducible case, I'll be sure to share!

In the meantime, I thought I'd share something interesting I came across: [Bug 18037 - infinite recursion (stack overflow) in regexec.c (sift_states_bkref->sift_states_backward->update_cur_sifted_state)](https://sourceware.org/bugzilla/show_bug.cgi?id=18037). I'm a C novice, but is it possible the regular expressions being used in this context are causing a recursive loop?

Many thanks!

**Comment by toddmazierski** [comment]
### Summary

Hi again, some progress to report: I have a repository where I can reproduce the crash consistently, and I'm able to reproduce it outside of Atom's UI. Unfortunately, it's proprietary code so I cannot share it.

@kevinsawicki, please let me know if you have any troubleshooting steps to share—I'm happy to fill up `repository.cc` with debugging statements. As a C/C++ novice, I'm not sure how to proceed any further. Thanks!

### From Atom

The crash is reproducible from the JavaScript console in Atom:

```javascript
git = require("git-utils")
// => Object {}
repository = git.open("/path/to/repo")
// => Repository {caseInsensitiveFs: true, submodules: Object}
repository.getDiffStats('issue_2586.php')
// => (crash)
```

This is the diff:

```diff
diff --git issue_2586.php issue_2586.php
index 852367d..b4f4984 100644
--- issue_2586.php
+++ issue_2586.php
@@ -1,2 +1 @@
 function someFunction() {
-}
```

**Please note:** strangely, if I take this same file/diff over to a different repository (ex. a brand new one), I **can't** reproduce the crash there.

### From the Git Node module test suite

I modified the `.getDiffStats(path)` test case in the Git Node module's (aka git-utils) [`git-spec.coffee`](https://github.com/atom/git-utils/blob/57177bcc710055b94838dcc49d08bcf9d7d2af1e/spec/git-spec.coffee#L345) file to point to the same local Git repository that causes Atom to crash:

```coffeescript
  describe '.getDiffStats(path)', ->
    beforeEach ->
      repoDirectory = '/path/to/repo' # <= same repository as above
      repo = git.open(repoDirectory)

    describe 'when the path is deleted', ->
      it 'returns of number of lines deleted', ->
        expect(repo.getDiffStats('issue_2586.php')).toEqual {added: 0, deleted: 1}
```

It causes the suite to fail in this way (it looks like a crash):

```
Running "shell:test" (shell) task
............................................Warning: Command failed: /bin/sh -c node node_modules/jasmine-focused/bin/jasmine-focused --captureExceptions --coffee spec/
 Use --force to continue.

Aborted due to warnings.
```

…if I modify `issue_2586.php` so Atom stops crashing (I'm not sure why this works):

```patch
diff --git issue_2586.php issue_2586.php
index 852367d..991216f 100644
--- issue_2586.php
+++ issue_2586.php
@@ -1,2 +1,4 @@
+// this won't cause a crash
+
 function someFunction() {
 }
```

The test suite appears to fail in a "normal" way:

```
Running "shell:test" (shell) task
............................................F........................

git
  .getDiffStats(path)
    when the path is deleted
      it returns of number of lines deleted
        Expected { added : 2, deleted : 0 } to equal { added : 0, deleted : 1 }. (spec/git-spec.coffee:428:62)


Finished in 5.21 seconds
69 tests, 168 assertions, 1 failure, 0 skipped
```

Hope this is useful! Thanks, again.

**Comment by viatsko** [comment]
Same bug here on a large repository :(

**Comment by viatsko** [comment]
I can confirm we are experiencing the same bug. I've patched libgit2 to exclude pattern_match to fix it, still trying to investigate the proper patch.

**Comment by carlosmn** [comment]
@toddmazierski can you share the repo/file which is causing this issue? If not, can you build libgit2 in debug mode and run it under valgrind or a debugger to get specifically where there is an issue?

**Comment by viatsko** [comment]
@carlosmn I'll do it too, will share shortly

**Comment by toddmazierski** [comment]
> …can you share the repo/file which is causing this issue?

Unfortunately, I can't (it's proprietary code).

> If not, can you build libgit2 in debug mode and run it under valgrind or a debugger to get specifically where there is an issue?

I'd be happy to, but I'm not well-versed enough in C to carry this out on my own. Is there a tutorial or documentation you could point me to?

**Comment by carlosmn** [comment]
I managed to get a minimal repro thanks to @viatsko; we're getting some odd output from regexec in that it's telling us the end-offset is 0 while the start-offset is 32 (for the particular string, of length 32) for a particular match, and the full-string match is 0, 0 which also seems rather nonsensical.

I haven't managed to reproduce this outside of libgit2 yet.

**Comment by carlosmn** [comment]
I believe I have found the cause of the issue. git-utils provides its own build system (gyp) for compiling libgit2. It however seems to include `deps/regex/` into the search path even outside of Windows.

What this means is that when libgit2 includes `<regex.h>`, it includes the GNU definition. This is bad because that means we have a definition of `regmatch_t` made out of two `int`s (4 bytes each). We however call to the Apple (system-provided, implicitly loaded) version of `regexec(3)`, which expects the Apple definition consisting of two `__off_t` (8 bytes each). So when we read the individual fields, we're actually reading the low and high 32-bit values of the actual value it wants to return. For extra lols, the system-provided `regexec(3)` writes outside of the stack we've allocated for our too-small `regmatch_t` so the actual crash happens *after* we return from the function, which explains some of the weird output when using printf-based debugging. I've no idea how this isn't crashing more often.

So the bug is in the git-utils build system.
